### DISCUSSION TITLE
ReferenceError는 어느 상황에서 발생하나요?

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#27] [변수]
- `ReferenceError`는 어느 상황에서 발생하나요?

**Comment by pipisebastian** [comment]
>ReferenceError는 어느 상황에서 발생하나요?

변수가 참조하는 메모리 공간을 찾지못할때와, 변수가 초기화되지 않은 상태에서 발생합니다.

그렇기 때문에 `var`키워드로 선언한 변수는, 선언과 동시에 메모리 할당과 초기화가 일어나기 때문에, `ReferenceError` 가 발생하지 않지만
`let`, `const` 키워드로 선언한 변수는 선언과 초기화사이에 TDZ라는 텀이 존재하기 때문에, 코드상에서 변수선언문에 도달하기전에 변수를 참조하게 되면, 초기화가 되지 않았기에 `ReferenceError` 가 발생합니다.

**Comment by publdaze** [comment]
> - `ReferenceError`는 어느 상황에서 발생하나요?
- let과 const에서 값을 선언하고, 할당하기 전 일시적 사각지대에서 발생
